collector/pg_index_size.go

```diff
@@ -0,0 +1,109 @@
+// Copyright 2023 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package collector
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const indexSizeSubsystem = "index_size"
+
+func init() {
+	registerCollector(indexSizeSubsystem, defaultDisabled, NewPGIndexSizeCollector)
+}
+
+type PGIndexSizeCollector struct {
+	log log.Logger
+}
+
+func NewPGIndexSizeCollector(config collectorConfig) (Collector, error) {
+	return &PGIndexSizeCollector{log: config.logger}, nil
+}
+
+var (
+	indexSizeDesc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, indexSizeSubsystem, "bytes"),
+		"Size of the index as per pg_table_size function",
+		[]string{"schemaname", "relname", "indexrelname"},
+		prometheus.Labels{},
+	)
+
+	indexSizeQuery = `
+	SELECT
+		schemaname,
+		tablename AS relname,
+		indexname AS indexrelname,
+		pg_class.relpages * 8192::bigint AS index_size
```

Comment:
We can't rely on 8192 being the block size. It's default, but could be changed. See references:
1. https://www.postgresql.org/docs/current/catalog-pg-class.html (see relpages)
2. https://www.postgresql.org/docs/current/runtime-config-preset.html (see block_size)

Reply:
ok so we have to select block size and multiply by that